rolling-shutter/medley/broker.go

```diff
@@ -0,0 +1,189 @@
+package medley
+
+import (
+	"sync"
+
+	"github.com/rs/zerolog/log"
+)
+
+// Broker allows to distribute a value to multiple
+// receiving channels, so that all channels receive the same value.
+// This is different to the Fan-Out pattern where multiple channels
+// are listening for a send on one channel, but only the first channel
+// to select the receive operation receives this value.
+//
+// The Broker allows to set whether the internal send operation is
+// blocking or non blocking. In the blocking case, all receiving
+// (subscribed) channels have to successively select the receive in order
+// for the broker to continue operation. So if one receiver blocks,
+// everything else will block as well.
+// In the non-blocking case, a subscribed channel that is blocking
+// while the published value is sent will miss the value.
+//
+// One has to be careful in when to select which case.
+// The blocking behavior can be partially subverted by increasing
+// the subscribers buffer size - but if a subscriber is blocked forever,
+// this will only work until this subscribers buffer is full.
+type Broker[T any] struct {
+	mux                  sync.RWMutex
+	stop                 chan struct{}
+	publish              chan T
+	subscribe            chan chan T
+	unsubscribe          chan chan T
+	initialSubscriptions []chan T
+	running              bool
+	nonBlockingSend      bool
+}
+
+func NewBroker[T any](nonBlocking bool) *Broker[T] {
+	return &Broker[T]{
+		stop:                 make(chan struct{}),
+		publish:              make(chan T, 1),
+		subscribe:            make(chan chan T),
+		unsubscribe:          make(chan chan T),
+		nonBlockingSend:      nonBlocking,
+		initialSubscriptions: []chan T{},
+		running:              false,
+	}
+}
+
+// Start starts the Broker's internal loop that processes
+// the subscription and publish operations.
+// initialSubscriptions will be registered
+// before the loop is started.
+// This is useful when the channels must not miss a published value
+// after the start or when the subscribers buffer size
+// should diverge from the default value `bufferSizeSubscriber`
+// set in the constructor.
+func (b *Broker[T]) Start() {
+	subscriptions := map[chan T]struct{}{}
+	for _, sub := range b.initialSubscriptions {
+		log.Debug().Int("num-subscribers", len(subscriptions)).Msg("subscribed initial subscription")
+		subscriptions[sub] = struct{}{}
+	}
+	started := make(chan struct{})
+	go func() {
+		b.mux.Lock()
+		b.running = true
+		b.mux.Unlock()
+		close(started)
+		for {
+			select {
+			case <-b.stop:
+				for msgCh := range subscriptions {
+					close(msgCh)
+					delete(subscriptions, msgCh)
+				}
+				b.mux.Lock()
+				b.running = false
+				b.mux.Unlock()
+				return
+			case msgCh := <-b.subscribe:
+				subscriptions[msgCh] = struct{}{}
+			case msgCh := <-b.unsubscribe:
+				delete(subscriptions, msgCh)
+				close(msgCh)
```

Comment:
I don't think it's the broker's responsibility to close the `msgCh` here as well as when stopping the broker. We  didn't create the channel, we should not close it.

Reply:
but isn't the [channel closing principle](https://stackoverflow.com/a/36613932/5053253) about who is sending / receiving values on it rather than who created it?
I think the sender to a channel is supposed to close it, so that on a receive we can check whether there will be more values coming, or whether we can disregard the channel:
```go
newVal, more := <- msgChan
if !more {
   msgChan = nil
}
```

Reply:
well, the case I'm thinking about is that one could use the same channel to subscribe to multiple brokers. It probably doesn't matter for our use case, so you can also keep the code as is. Though in principle I still do think the we should either manage lifetimes of individual channels completely inside or completely outside the broker.